File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
from collections import deque

def solution(x, y, n):
    q = deque([x])
    visited = [0] * 1000001
    while q:
        k = q.popleft()
        if k == y:
            return visited[k]
        for v in [k + n, k * 2, k * 3]:
            if not (0 <= v < 1000001) or visited[v]:
                continue
            visited[v] = visited[k] + 1
            q.append(v)
    return -1

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.*;

class Solution {
    public int solution(int x, int y, int n) {
        Deque<Integer> q = new ArrayDeque<>();
        int[] visited = new int[1000001];
        
        q.offer(x);
        while (!q.isEmpty()) {
            int k = q.poll();
            if (k == y) return visited[k];
            for (long v : new long[]{k + n, k * 2, k * 3}) {
                if (v > 1000000 || visited[(int) v] != 0) continue;
                visited[(int) v] = visited[k] + 1;
                q.offer((int) v);
            }
        }
        return -1;
    }
}